File: nvpl_blas/f77/dgemmt.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DGEMMT
!
!******************************************************************************
      program   DGEMMT_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

      character*1     transa, transb, uplo
      parameter       (transa='T', transb='N', uplo='L')
      integer         n, k, i, j
      integer         lda, ldb, ldc
      parameter       (n=2, k=2,
     &                 lda=2, ldb=2, ldc=2)
      double precision alpha, beta
      parameter       (alpha=0.6d0, beta=0.4d0)
      double precision a(lda,k), b(ldb,n), c(ldc,n)
*       External Subroutines
      external        print_dmatrix, fill_dmatrix
*
*      Executable Statements
*
      print*
      print 99, 'Example: DGEMMT performs of the matrix-matrix' //
     &          ' operations with general matrices and add the' //
     &          ' result to the upper or lower part of a' //
     &          ' scalar-matrix product.'
*
*       Print input data
      print*
      print 100, n, k, lda, ldb, ldc, transa, transb, uplo
      print 101, alpha, beta
      call fill_dmatrix(a,n,k,lda,'F',transa)
      call fill_dmatrix(b,k,n,ldb,'F',transb)
      call fill_dmatrix(c,n,n,ldc,uplo,'N')
      call print_dmatrix(a,n,k,lda,'A')
      call print_dmatrix(b,k,n,ldb,'B')
      call print_dmatrix(c,n,n,ldc,'C')
*
*      Call DGEMMT subroutine
      call DGEMMT(uplo,transa,transb,n,k,alpha,a,lda,b,ldb,beta,c,ldc)
*
      print*
      print 102, 'After DGEMMT operation:'
      call print_dmatrix(c,n,n,ldc,'C')

      stop
 99   format(A)
 100  format('#### args: n=',i1,', k=',i1,', lda=',i1,
     &       ', ldb=',i1,', ldc=',i1,', transa=',a1,', transb=',a1,
     &       ', uplo=',a1)
 101  format(11x,'alpha=',f3.1,', beta=',f3.1)
 102  format(A)
      end

File: nvpl_blas/c/dgemmt.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_dgemmt
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t K = 2;
    nvpl_int_t lda = 2;
    nvpl_int_t ldb = 2;
    nvpl_int_t ldc = 2;
    double alpha = 0.6;
    double beta = 0.4;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_UPLO uplo = CblasLower;
    enum CBLAS_TRANSPOSE transA = CblasTrans;
    enum CBLAS_TRANSPOSE transB = CblasNoTrans;
    double * A = NULL;
    double * B = NULL;
    double * C = NULL;

    printf("\nExample: cblas_dgemmt for symmetric matrix-matrix multiplication\n\n");
    printf("#### args: m=%" PRId64 ", k=%" PRId64 ", lda=%" PRId64 ", ldb=%" PRId64 ", ldc=%" PRId64 ", "
            "order=%c, uplo=%c, transA=%c, transB=%c\n", (int64_t)N, (int64_t)K, (int64_t)lda, (int64_t)ldb,
            (int64_t)ldc, order_to_char(order), uplo_to_char(uplo), 
            transpose_to_char(transA), transpose_to_char(transB));
    printf("           alpha=%g, beta=%g\n", alpha, beta);

    nvpl_int_t rowsA = (transA == CblasNoTrans) ? N : K;
    nvpl_int_t colsA = (transA == CblasNoTrans) ? K : N;
    nvpl_int_t rowsB = (transB == CblasNoTrans) ? K : N;
    nvpl_int_t colsB = (transB == CblasNoTrans) ? N : K;
    nvpl_int_t rowsC = N;
    nvpl_int_t colsC = N;

    // allocate memory
    A = (double *)malloc(lda * colsA * sizeof(double));
    B = (double *)malloc(ldb * colsB * sizeof(double));
    C = (double *)malloc(ldc * colsC * sizeof(double));

    // fill data
    fill_dmatrix(A, rowsA, colsA, lda, order, Full, CblasNonUnit);
    fill_dmatrix(B, rowsB, colsB, ldb, order, Full, CblasNonUnit);
    fill_dmatrix(C, rowsC, colsC, ldc, order, UPLO_TO_FILL_MODE(uplo), CblasNonUnit);

    // print input data
    print_dmatrix(A, rowsA, colsA, lda, "A", order);
    print_dmatrix(B, rowsB, colsB, ldb, "B", order);
    print_dmatrix(C, rowsC, colsC, ldc, "C", order);

    // call cblas_dgemmt
    cblas_dgemmt(order, uplo, transA, transB, N, K, alpha, A, lda, B, ldb, beta, C, ldc);

    // print result
    printf("\nAfter dgemmt operation:\n");
    print_dmatrix(C, rowsC, colsC, ldc, "C", order);

    // release memory
    free(A);
    free(B);
    free(C);
    return 0;
}
